src/core/qgsnullpainterdevice.h

```diff
@@ -0,0 +1,109 @@
+/***************************************************************************
+  qgsnullpainterdevice.h
+  --------------------------------------
+  Date                 : December 2021
+  Copyright            : (C) 2013 by Mathieu Pellerin
+  Email                : nirvn dot asia at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSNULLPAINTERDEVICE_H
+#define QGSNULLPAINTERDEVICE_H
+
+#define SIP_NO_FILE
+
+#include "qgis_core.h"
+
+#include <QPaintDevice>
+#include <QPaintEngine>
+
+
+class QgsNullPaintEngine: public QPaintEngine
+{
+
+  public:
+
+    QgsNullPaintEngine() : QPaintEngine( QPaintEngine::AllFeatures ) {};
+
+    bool begin( QPaintDevice * ) override { return true; };
+    bool end() override { return true; };
+    QPaintEngine::Type type() const override { return QPaintEngine::User; };
+    void updateState( const QPaintEngineState & ) override { return; };
+    void drawPixmap( const QRectF &, const QPixmap &, const QRectF & ) override { return; };
```

Comment:
I think for performance this should implement ALL the virtual methods -- otherwise the default fallbacks will be doing unwanted work (eg drawing to pixmap, then calling the noop drawPixmap call)

Reply:
@nyalldawson , done.